src/client/src/pricingOperations.ts

```diff
@@ -1,66 +1,98 @@
+
+const MS_FOR_LAST_PRICE_TO_BECOME_STALE = 6000
 
-const getCurrencyPairs = (symbols: Array<string>, pricingService$: any) => {
-  return Observable.from(symbols).mergeMap(symbol => pricingService$.getSpotPriceStream({ symbol }))
+interface PricesBySymbol {
+  [symbol: string]: SpotPriceTick
 }
 
-const accumulatePrices = (acc, tick, index) => {
+type State = PricesBySymbol
+
+const reducePrices = (acc: PricesBySymbol, tick: SpotPriceTick): PricesBySymbol => {
   return {
     ...acc,
     [tick.symbol]: tick,
   }
 }
 
+const getSymbol = (currency): string => currency.currencyPair.symbol
+const getSymbols = (action): string[] => action.payload.currencyPairUpdates.map(getSymbol)
+// returns a stream that emits an array containing the currencyPairSymbols like this: `["EURUSD", ...]`
+const getSymbolArray$ = action$ => action$.ofType(REF_ACTION_TYPES.REFERENCE_SERVICE).map(getSymbols)
+// returns a stream that emits each currencyPairSymbol individually. Example: `EURUSD`
+const getSymbol$ = action$ => getSymbolArray$(action$).mergeMap((symbols: string[]) => Observable.from(symbols))
+
+const publishPriceToOpenFin = openFin => price => openFin.publishPrice(price)
+const addRatePrecisionToPrice = referenceDataService => price => {
+  return {
+    ...price,
+    ratePrecision: referenceDataService.getCurrencyPair(price.symbol).ratePrecision
+  }
+}
+
+// Creates a stream that only start emitting new prices after `REFERENCE_SERVICE` action is received.
+// Also only emits for symbols specified in `REFERENCE_SERVICE`.
+const priceForReferenceServiceSymbols$ = (action$, pricingService$) => {
+  // For each symbol
+  return getSymbol$(action$)
+  // get a new price stream for that symbol
+    .mergeMap(symbol => pricingService$.getSpotPriceStream({symbol}))
+}
+
+// @todo: split in different files
 export const pricingServiceEpic = (pricingService$, openFin, referenceDataService) => {
+  const stalePriceEpic = action$ => {
+    // For each symbol
+    return getSymbol$(action$)
+    // creates a new price stream and will wait to debounce
+      .mergeMap(symbol => pricingService$.getSpotPriceStream({symbol}).debounceTime(MS_FOR_LAST_PRICE_TO_BECOME_STALE))
+      // when debounces, it creates an action
+      .map(createStalePriceAction)
+  }
 
-  const pricingStream$ = action$ => action$.ofType(REF_ACTION_TYPES.REFERENCE_SERVICE)
-    .map(action => action.payload.currencyPairUpdates.map(currency => currency.currencyPair.symbol))
-
-  function stalePriceCheck(action$) {
-    return pricingStream$(action$)
-      .mergeMap((symbols: Array<string>) => {
-        return Observable.from(symbols)
-          .mergeMap(symbol => pricingService$.getSpotPriceStream({ symbol })
-            .debounce(() => Observable.interval(PRICE_STALE_AFTER_X_IN_MS))
-            .map(stalePricing))
-      })
+  const publishPriceToOpenFinEpic = action$ => {
+    // @todo: `publishPriceToOpenFin` should be in the middleware to play nicely with the epics.
```

Comment:
are you going to address the 'todo' items in this PR?

Reply:
No because that is a big change. I created this issues here: https://github.com/AdaptiveConsulting/ReactiveTraderCloud/issues/637

Reply:
ok, could you please remove the comments?